sycl/include/sycl/ext/oneapi/bfloat16.hpp

```diff
@@ -322,6 +326,309 @@ inline bfloat16 bitsToBfloat16(const Bfloat16StorageT Value) {
+    uint16_t fp64_exp = static_cast<uint16_t>((u64_val >> 52) & 0x7FF);
+    uint64_t fp64_mant = (u64_val & 0xF'FFFF'FFFF'FFFF);
+    uint16_t bf16_mant;
+    // handling +/-infinity and NAN for double input
+    if (fp64_exp == 0x7FF) {
+      if (!fp64_mant) {
+        return bf16_sign ? 0xFF80 : 0x7F80;
+      } else {
+        // returns a quiet NaN
+        return 0x7FC0;
+      }
+    }
+
+    // Subnormal double precision is converted to 0
+    if (fp64_exp == 0) {
+      return bf16_sign ? 0x8000 : 0x0;
+    }
+
+    fp64_exp -= 1023;
+    // handling overflow, convert to +/-infinity
+    if (static_cast<int16_t>(fp64_exp) > 127) {
+      return bf16_sign ? 0xFF80 : 0x7F80;
+    }
+
+    // handling underflow
+    if (static_cast<int16_t>(fp64_exp) < -133) {
+      return bf16_sign ? 0x8000 : 0x0;
+    }
+
+    //-133 <= fp64_exp <= 127, 1.signicand * 2^fp64_exp
+    // For these numbers, they are NOT subnormal double-precision numbers but
+    // will turn into subnormal when converting to bfloat16
+    uint64_t discard_bits;
+    if (static_cast<int16_t>(fp64_exp) < -126) {
+      fp64_mant |= 0x10'0000'0000'0000;
+      fp64_mant >>= -126 - static_cast<int16_t>(fp64_exp) - 1;
+      discard_bits = fp64_mant & 0x3FFF'FFFF'FFFF;
+      bf16_mant = static_cast<uint16_t>(fp64_mant >> 46);
+      if (discard_bits > 0x2000'0000'0000 ||
+          ((discard_bits == 0x2000'0000'0000) && ((bf16_mant & 0x1) == 0x1)))
+        bf16_mant += 1;
+      fp64_exp = 0;
+      if (bf16_mant == 0x80) {
+        bf16_mant = 0;
+        fp64_exp = 1;
+      }
+      return (bf16_sign << 15) | (fp64_exp << 7) | bf16_mant;
+    }
+
+    // For normal value, discard 45 bits from mantissa
+    discard_bits = fp64_mant & 0x1FFF'FFFF'FFFF;
+    bf16_mant = static_cast<uint16_t>(fp64_mant >> 45);
+    if (discard_bits > 0x1000'0000'0000 ||
+        ((discard_bits == 0x1000'0000'0000) && ((bf16_mant & 0x1) == 0x1)))
+      bf16_mant += 1;
+
+    if (bf16_mant == 0x80) {
+      if (fp64_exp != 127) {
+        bf16_mant = 0;
+        fp64_exp++;
+      } else {
+        return bf16_sign ? 0xFF80 : 0x7F80;
+      }
+    }
+    fp64_exp += 127;
+
+    return (bf16_sign << 15) | (fp64_exp << 7) | bf16_mant;
+  }
+
+public:
+  template <typename Ty, int rm>
+  static bfloat16 getBfloat16WithRoundingMode(const Ty &a) {
+
+    if (!a)
+      return bfloat16{0.0f};
+
+    constexpr SYCLRoundingMode roundingMode = static_cast<SYCLRoundingMode>(rm);
+
+    // Float.
+    if constexpr (std::is_same_v<Ty, float>) {
+      return getBFloat16FromFloatWithRoundingMode(a, roundingMode);
+    }
+    // Double.
+    else if constexpr (std::is_same_v<Ty, double>) {
+      static_assert(
+          roundingMode == SYCLRoundingMode::automatic ||
+              roundingMode == SYCLRoundingMode::rte,
+          "Only automatic/RTE rounding mode is supported for double type.");
+      return getBFloat16FromDoubleWithRoundingMode(a, roundingMode);
+    }
```

Comment:
is there a possibility of other floating types besides float and double?  Half?  Should there be a `! std::is_integral_v<Ty>` clause for the future?

Reply:
Nice catch. I've added the clause for `half` as well.